File: mycodes/875-KokoEatingBananas.java
//Time 86%
//Space 95%

class Solution {
    public int minEatingSpeed(int[] piles, int H) {
        //越慢越好->K越小越好
        //可以运用二分查找，low为1，high为piles中的最大值
        //对于每一个K，若消耗的时间大于H，则K小了，否则可能比要求的最小K大了
        //时间复杂度O(nlogm) m为香蕉最大数，n为串数。
        if (H < piles.length) return -1;
        int K, low = 1, high = getMax(piles);
        while (low < high) {
            K = low + (high - low) / 2;
            if (getEatTime(piles, K) > H) low = K + 1;
            else high = K;
        }
        return high;
    }
    
    public int getMax(int[] piles) {
        int max = Integer.MIN_VALUE;
        for (int v : piles) max = Math.max(max, v);
        return max;
    }
    
    public int getEatTime(int[] piles, int K) {
        int res = 0;
        for (int v : piles) res += (v - 1) / K + 1;
        return res;
    }
}

File: mycodes/875-KokoEatingBananas.go
func minEatingSpeed(piles []int, h int) int {
  l, hi := 1, getMaxSpeed(piles)
  for l < hi {
    m := l + (hi-l)/2
    hour := totalHour(piles, m)
    if hour <= h {
      hi = m
    } else {
      l = m + 1
    }
  }
  return hi
}

func totalHour(piles []int, s int) int {
  res := 0
  for _, p := range piles {
    h := p / s
    if p % s > 0 {
      h++
    }
    res += h
  }
  return res
}

func getMaxSpeed(piles []int) int {
  res := 0
  for _, p := range piles {
    res = max(res, p)
  }
  return res
}
